### ISSUE TITLE
JSON with a database storage backend

### ISSUE THREAD

**Issue by JohannesLichtenberger** [sebbae/brackit#5]
Hey Sebastian, sorry to bother you all the time, but maybe you can also give a good tip how you'd implement a binding for a native JSON storage backend. I'm currently finishing the work on a temporal JSON storage layer besides XML for my open source project (http://sirix.io) and would like to also use Brackit. Maybe I could also implement the interfaces you already provide (should be basically the same as for my XDM node transactions) and then implement some functions for opening a JSON resource...

Basically my JSON node layer looks like the one in Marklogic:
https://docs.marklogic.com/guide/app-dev/json

**Comment by sebbae** [comment]
I'm not sure if I understand you correctly. For the current JSON capabilities of Brackit there are interfaces like org.brackit.xquery.xdm.Record and org.brackit.xquery.xdm.Array etc. Did you mean them?

If you have a concrete storage backend for JSON it should be possible to start with implementing these interfaces just like BrackitDB implements the (XML) Node interface with TXNode and BracketNode.

Is this what you've been looking for?

BR, Sebastian

**Comment by JohannesLichtenberger** [comment]
Yes, but I thought about something like a Node interface which encapsulates both Arrays, Records and the JSON value types. But yeah, you might be right that I simply can implement those two interfaces and an additional database store... will keep you updated :-)

**Comment by JohannesLichtenberger** [comment]
Hi Sebastian, haven't had the time to look at it until now, but shouldn't I introduce a JsonStore, JsonCollection and a JsonNodeFactory or extend the current NodeFactory? I guess that it be easier to write some functions like json:doc, json:collection... And I could simply abstract all the stuff, as my JSON and XML APIs basically are the same with some generic parameters replaced :-) that said I've read that one of your students once wrote a MongoDB binding and several more (Cassandra!?) maybe just with the Array and Record interfaces.

**Comment by JohannesLichtenberger** [comment]
I've added structured items to the type hierarchy of Brackit, which nodes and arrays/objects (or records) simply extend. This way we can easily abstract all store/collection... stuff :-) (so it's compatible to JSONiq).

**Comment by sebbae** [comment]
Hi Johannes,

Brackit was basically designed around the [XQuery and XPath Data Model (XDM)](https://www.w3.org/TR/xpath-datamodel-31/#types-hierarchy) which - put simply - states that XQuery operates on items (nodes, functions, atomic values) and sequences of items.

Items carry additional typing information like xs:double, xs:untyped etc. but this is a typing dimension which is important at runtime and not for the design of the query engine itself.

Conceptually, I don't see an obvious connection between a JSON element and a "node". They are different "things".

For supporting JSON, I therefore chose the straightforward solution and simply introduced `Record` and `Array` as additional item types in brackit's XDM structure side-by-side to the existing ones like `Node`, `Function`, and `Atomic`.  

From there, the rest fell in place quite naturally, as I "only" had to extend the parser and compiler to support additional expression types on these new item types and provide a few utility functions in the `bit` and `json` namespaces. :-)

For processing JSON via a native JSON storage backed I would assume it were sufficient for the start to introduce XQuery functions which read from and write to the JSON store. Just like `fn:doc()` is conceptually nothing more like a function which returns an XML document node from an underlying, but not further specified XML store.

So, if you have such functions which translate between the on disk representation and the runtime representation as `Record` or `Array` item you could be already good to go.

Of course, for boosting query performance, e.g. for pushing down a filter predicate to the native store additional work needs to be done. However, this would then be a job for the query optimizer!

If you can match an expression, i.e. a part of the query AST, which maps to an operation or sequence of operations which can be natively supported by your JSON store, you can rewrite the AST accordingly and compile it to an efficient native operation of your JSON backend.

BrackitDB does this kind of optimized processing for example for sequences of multiple child axis steps. If a query has a sequence like `$x/a/b/c` it replaces the canonic iterative traversal of the DOM into a more efficient scan operation for nodes on the store.

I really does not matter if you want to leverage raw scan performance of your storage or index structures or even a potential ability to execute a specific type of join natively. You'll only need to match the opportunity to do so in the query AST and rewrite that part of the query to a new kind of expression type which can be compiled and returns a sequence of `Record`'s, `Array`'s or whatever is approrpiate.

I hope this lengthy explanations were helpful!

Best regards,
Sebastian

**Comment by sebbae** [comment]
I just realized that "map" and "array", which are basically equivalent to JSON's record and array, are actually now officially a part of XDM. They have been added 2014 according to the changelog.

So, I think I was ahead of time with my design choice. ;-)

**Comment by JohannesLichtenberger** [comment]
Hi Sebastian, yeah it's working I guess.

Basically I'm now using this type hierarchy:

http://www.jsoniq.org/docs/JSONiqExtensionToXQuery/html-single/index.html#idm38012096

I also changed the parser slightly:

https://github.com/sirixdb/brackit/commit/3df529809e61e5f1894291cd21aebcb1890c9fd5

Now we can do =>foo[[0]]=>bar that is we're able to mix JSON deref expressions and array index accesses. I'd like to support array slices, too, for instance [[0:5]] to retrieve the first 6 array elements, but I just did a quick test. Have to change the parser slightly again. But in your interface it's already defined.

I also basically added a load, store, doc, collection, indexing and a bunch more of functions :-)

Just want to play around a bit and write more integration tests and hopefully release a new version of Sirix in a couple of days :-) basically I also added a lot of functions and parameters to take the temporal aspects of Sirix into account, so you can query objects, arrays or XML nodes in any revision and get future, past, all-time, next, previous and so on versions.

I guess there's plenty of opportunity to rewrite the query plan and get better performance, but I'm already happy now having an XML- as well as JSON storage layer side by side + sophisticated querying capabilities through Brackit and temporal enhancements :) from there on I can optimize stuff or maybe first decide and implement how to replicate and partition a database ;)

**Comment by JohannesLichtenberger** [comment]
That said I'm not sure why I used the JSONiq extended XQuery data model, even if it was just a few weeks ago :-D I think it was that I can use a few of my custom functions for both JSON items (arrays/objects) and XML nodes.

I also introduced a JsonStore and a JsonCollection and so on analogous to the XML node stuff.

Basic querying, even temporal stuff should work already :-)
